File: tests/analysis.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { auditEvidence } from "../src/audit.js";
import { diffGraphs } from "../src/diff.js";
import { buildGraph } from "../src/graph.js";
import { buildProvenance } from "../src/provenance.js";
import { renderMarkdownReport } from "../src/report.js";
import type { DiscoveredItem } from "../src/types.js";

function item(overrides: Partial<DiscoveredItem> & Pick<DiscoveredItem, "id" | "kind" | "sourcePath" | "scope" | "precedence">): DiscoveredItem {
  return {
    agent: "claude-code",
    parser: "json",
    sensitivity: "command_config",
    contentPolicy: "structured_safe_fields_only",
    restorePolicy: "not_supported",
    captureStatus: "captured",
    confidence: "high",
    ...overrides
  };
}

describe("analysis modules", () => {
  it("builds a graph where project policy overrides user policy and provenance stays JSON-friendly", () => {
    const evidence = [
      item({
        id: "user-permission-bash",
        kind: "permission",
        sourcePath: "~/.claude/settings.json",
        scope: "user",
        precedence: 10,
        name: "Bash(git status)",
        value: { action: "allow", rule: "Bash(git status)" }
      }),
      item({
        id: "project-permission-bash",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(git status)",
        value: { action: "deny", rule: "Bash(git status)" }
      })
    ];

    const graph = buildGraph(evidence);
    const userNode = graph.find((node) => node.evidenceId === "user-permission-bash");
    const projectNode = graph.find((node) => node.evidenceId === "project-permission-bash");

    assert.ok(userNode);
    assert.ok(projectNode);
    assert.equal(userNode.overriddenBy, projectNode.id);
    assert.deepEqual(projectNode.effectiveValue, { action: "deny", rule: "Bash(git status)" });

    const provenance = buildProvenance(graph, evidence);
    assert.deepEqual(provenance.find((entry) => entry.evidenceId === "project-permission-bash"), {
      nodeId: projectNode.id,
      evidenceId: "project-permission-bash",
      sourcePath: ".claude/settings.json",
      scope: "project",
      precedence: 40,
      confidence: "high",
      captureStatus: "captured"
    });
  });

  it("audits project override, permission wildcard, parse failure, symlink, and omitted secret-like values", () => {
    const evidence = [
      item({
        id: "user-policy",
        kind: "permission",
        sourcePath: "~/.claude/settings.json",
        scope: "user",
        precedence: 10,
        name: "Bash(git status)",
        value: { action: "allow", rule: "Bash(git status)" }
      }),
      item({
        id: "project-policy",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(git status)",
        value: { action: "deny", rule: "Bash(git status)" }
      }),
      item({
        id: "project-wildcard",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(*)",
        value: { action: "allow", rule: "Bash(*)" }
      }),
      item({
        id: "bad-json",
        kind: "agent_config",
        sourcePath: ".mcp.json",
        scope: "project",
        precedence: 40,
        captureStatus: "parse_failed",
        metadata: { error: "Unexpected token" }
      }),
      item({
        id: "skipped-link",
        kind: "symlink",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        captureStatus: "omitted",
        metadata: { skipped: true, target: "../private/settings.json" }
      }),
      item({
        id: "env-token",
        kind: "env_key",
        sourcePath: ".env",
        scope: "project",
        precedence: 40,
        captureStatus: "omitted",
        name: "OPENAI_API_KEY",
        metadata: { secretLike: true }
      })
    ];

    const findings = auditEvidence(evidence, buildGraph(evidence));
    const codes = findings.map((finding) => finding.code).sort();

    assert.deepEqual(codes, [
      "PARSE_FAILED",
      "PERMISSION_WILDCARD_ADDED",
      "PROJECT_OVERRIDES_USER_POLICY",
      "SECRET_LIKE_VALUE_OMITTED",
      "SYMLINK_SKIPPED"
    ]);
    for (const finding of findings) {
      assert.ok(finding.problem);
      assert.ok(finding.cause);
      assert.ok(finding.fix);
      assert.ok(finding.path);
      assert.ok(finding.evidenceId);
    }
  });

  it("diffs semantic MCP changes and raw source changes", () => {
    const baselineGraph = buildGraph([
      item({
        id: "github-old",
        kind: "mcp_server",
        sourcePath: ".mcp.json",
        scope: "project",
        precedence: 40,
        name: "github",
        value: { transport: "stdio", command: "mcp-github", args: ["--read-only"] },
        checksum: "old"
      })
    ]);
    const currentGraph = buildGraph([
      item({
        id: "github-new",
        kind: "mcp_server",
        sourcePath: ".mcp.json",
        scope: "project",
        precedence: 40,
        name: "github",
        value: { transport: "http", url: "https://mcp.example.com/github" },
        checksum: "new"
      })
    ]);

    const diff = diffGraphs(baselineGraph, currentGraph);

    assert.deepEqual(diff.rawSourceChanges, [
      {
        sourcePath: ".mcp.json",
        beforeEvidenceId: "github-old",
        afterEvidenceId: "github-new",
        beforeChecksum: undefined,
        afterChecksum: undefined,
        status: "changed"
      }
    ]);
    assert.equal(diff.semanticChanges.length, 1);
    assert.equal(diff.semanticChanges[0].code, "MCP_CHANGED");
    assert.equal(diff.semanticChanges[0].entityName, "github");
    assert.deepEqual(diff.semanticChanges[0].details.changedFields.sort(), ["command", "transport", "urlHost"]);
  });

  it("diffs setup inventory changes used for deterministic save titles", () => {
    const baselineGraph = buildGraph([
      item({
        id: "instructions-old",
        kind: "agent_instruction",
        sourcePath: "AGENTS.md",
        scope: "project",
        precedence: 40,
        name: "AGENTS.md",
        value: { checksum: "old" },
        checksum: "old"
      }),
      item({
        id: "skill-old",
        kind: "skill",
        sourcePath: ".claude/skills/legacy-review/SKILL.md",
        scope: "project",
        precedence: 40,
        name: "legacy-review",
        value: { installed: true },
        checksum: "legacy"
      }),
      item({
        id: "hook-old",
        kind: "hook",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "pre-tool-use",
        value: { command: "old-hook" },
        checksum: "hook-old"
      }),
      item({
        id: "permission-old",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(git status)",
        value: { action: "allow", rule: "Bash(git status)" },
        checksum: "permission-old"
      }),
      item({
        id: "permission-removed",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(bun test)",
        value: { action: "allow", rule: "Bash(bun test)" },
        checksum: "permission-removed"
      })
    ]);
    const currentGraph = buildGraph([
      item({
        id: "instructions-new",
        kind: "agent_instruction",
        sourcePath: "AGENTS.md",
        scope: "project",
        precedence: 40,
        name: "AGENTS.md",
        value: { checksum: "new" },
        checksum: "new"
      }),
      item({
        id: "skill-new",
        kind: "skill",
        sourcePath: ".claude/skills/react-review/SKILL.md",
        scope: "project",
        precedence: 40,
        name: "react-review",
        value: { installed: true },
        checksum: "skill"
      }),
      item({
        id: "hook-new",
        kind: "hook",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "pre-tool-use",
        value: { command: "new-hook" },
        checksum: "hook-new"
      }),
      item({
        id: "hook-added",
        kind: "hook",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "post-tool-use",
        value: { command: "notify" },
        checksum: "hook-added"
      }),
      item({
        id: "permission-new",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(git status)",
        value: { action: "deny", rule: "Bash(git status)" },
        checksum: "permission-new"
      }),
      item({
        id: "permission-added",
        kind: "permission",
        sourcePath: ".claude/settings.json",
        scope: "project",
        precedence: 40,
        name: "Bash(bun run build)",
        value: { action: "allow", rule: "Bash(bun run build)" },
        checksum: "permission-added"
      })
    ]);

    const diff = diffGraphs(baselineGraph, currentGraph);

    assert.deepEqual(
      diff.semanticChanges.map((change) => change.code).sort(),
      [
        "HOOK_ADDED",
        "HOOK_CHANGED",
        "INSTRUCTION_CHANGED",
        "PERMISSION_CHANGED",
        "PERMISSION_CHANGED",
        "PERMISSION_CHANGED",
        "SKILL_ADDED",
        "SKILL_REMOVED"
      ]
    );
    assert.equal(
      diff.semanticChanges.some((change) => change.code === "PERMISSION_CHANGED" && change.details.removed === true),
      true
    );
  });

  it("renders markdown reports with detected agents, blind spots, findings, provenance, and next command", () => {
    const evidence = [
      item({
        id: "unsupported-codex",
        agent: "codex",
        kind: "unsupported",
        sourcePath: ".codex/config.toml",
        scope: "project",
        precedence: 30,
        captureStatus: "unsupported",
        confidence: "medium",
        metadata: { state: "present" }
      })
    ];
    const graph = buildGraph(evidence);
    const findings = auditEvidence(evidence, graph);
    const provenance = buildProvenance(graph, evidence);

    const markdown = renderMarkdownReport({
      snapshotName: "current",
      trust: { readOnly: true, network: "disabled", commandsExecuted: 0 },
      evidence,
      graph,
      findings,
      provenance,
      blindSpots: ["Remote MCP server behavior cannot be captured", "Raw env values are omitted by policy"]
    });

    assert.match(markdown, /^# hem report: current/m);
    assert.match(markdown, /## Detected agents/);
    assert.match(markdown, /Codex\s+project state found/);
    assert.match(markdown, /## High-signal findings/);
    assert.match(markdown, /UNSUPPORTED_AGENT_STATE/);
    assert.doesNotMatch(markdown, /hem v0\.1/);
    assert.match(markdown, /cannot yet be interpreted by hem/);
    assert.match(markdown, /## Blind spots/);
    assert.match(markdown, /Remote MCP server behavior cannot be captured/);
    assert.match(markdown, /## Reproducibility gaps/);
    assert.match(markdown, /unsupported: 1/);
    assert.match(markdown, /## Provenance/);
    assert.match(markdown, /unsupported-codex/);
    assert.match(markdown, /hem snapshot create --name baseline --metadata-only --project \./);
  });
});

File: crates/hem-core/tests/analysis_test.rs
use hem_core::{
    audit_evidence, build_graph, build_provenance, diff_graphs, AgentId, CaptureStatus,
    DiscoveredItem, EvidenceConfidence, EvidenceKind, EvidenceParser, EvidenceScope,
    RestorePolicy, SemanticChangeCode,
};

fn item(
    id: &str,
    kind: EvidenceKind,
    source_path: &str,
    scope: EvidenceScope,
    precedence: u32,
    overrides: ItemOverrides,
) -> DiscoveredItem {
    DiscoveredItem {
        id: id.to_string(),
        agent: overrides.agent.unwrap_or(AgentId::ClaudeCode),
        kind,
        source_path: source_path.to_string(),
        scope,
        precedence,
        parser: EvidenceParser::Json,
        sensitivity: "command_config".to_string(),
        content_policy: "structured_safe_fields_only".to_string(),
        restore_policy: RestorePolicy::NotSupported,
        capture_status: overrides.capture_status.unwrap_or(CaptureStatus::Captured),
        confidence: EvidenceConfidence::High,
        name: overrides.name,
        value: overrides.value,
        checksum: overrides.checksum,
        metadata: overrides.metadata,
    }
}

struct ItemOverrides {
    agent: Option<AgentId>,
    name: Option<String>,
    value: Option<serde_json::Value>,
    checksum: Option<String>,
    capture_status: Option<CaptureStatus>,
    metadata: Option<serde_json::Value>,
}

impl ItemOverrides {
    fn with_name(mut self, name: &str) -> Self {
        self.name = Some(name.to_string());
        self
    }

    fn with_value(mut self, value: serde_json::Value) -> Self {
        self.value = Some(value);
        self
    }

    fn with_checksum(mut self, checksum: &str) -> Self {
        self.checksum = Some(checksum.to_string());
        self
    }

    fn with_capture_status(mut self, status: CaptureStatus) -> Self {
        self.capture_status = Some(status);
        self
    }

    fn with_metadata(mut self, metadata: serde_json::Value) -> Self {
        self.metadata = Some(metadata);
        self
    }
}

impl Default for ItemOverrides {
    fn default() -> Self {
        Self {
            agent: None,
            name: None,
            value: None,
            checksum: None,
            capture_status: None,
            metadata: None,
        }
    }
}

#[test]
fn builds_graph_with_project_override_and_provenance() {
    let evidence = vec![
        item(
            "user-permission-bash",
            EvidenceKind::Permission,
            "~/.claude/settings.json",
            EvidenceScope::User,
            10,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(git status)"})),
        ),
        item(
            "project-permission-bash",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "deny", "rule": "Bash(git status)"})),
        ),
    ];

    let graph = build_graph(&evidence);
    let user_node = graph
        .iter()
        .find(|node| node.evidence_id == "user-permission-bash")
        .expect("user node");
    let project_node = graph
        .iter()
        .find(|node| node.evidence_id == "project-permission-bash")
        .expect("project node");

    assert_eq!(user_node.overridden_by.as_deref(), Some(project_node.id.as_str()));
    assert_eq!(
        project_node.effective_value,
        serde_json::json!({"action": "deny", "rule": "Bash(git status)"})
    );

    let provenance = build_provenance(&graph, &evidence);
    assert_eq!(
        provenance
            .iter()
            .find(|entry| entry.evidence_id == "project-permission-bash"),
        Some(&hem_core::ProvenanceEntry {
            node_id: project_node.id.clone(),
            evidence_id: "project-permission-bash".to_string(),
            source_path: ".claude/settings.json".to_string(),
            scope: EvidenceScope::Project,
            precedence: 40,
            confidence: EvidenceConfidence::High,
            capture_status: CaptureStatus::Captured,
        })
    );
}

#[test]
fn audits_project_override_wildcard_parse_failure_symlink_and_secret_like_values() {
    let evidence = vec![
        item(
            "user-policy",
            EvidenceKind::Permission,
            "~/.claude/settings.json",
            EvidenceScope::User,
            10,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(git status)"})),
        ),
        item(
            "project-policy",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "deny", "rule": "Bash(git status)"})),
        ),
        item(
            "project-wildcard",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(*)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(*)"})),
        ),
        item(
            "bad-json",
            EvidenceKind::AgentConfig,
            ".mcp.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_capture_status(CaptureStatus::ParseFailed)
                .with_metadata(serde_json::json!({"error": "Unexpected token"})),
        ),
        item(
            "skipped-link",
            EvidenceKind::Symlink,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_capture_status(CaptureStatus::Omitted)
                .with_metadata(serde_json::json!({"skipped": true, "target": "../private/settings.json"})),
        ),
        item(
            "env-token",
            EvidenceKind::EnvKey,
            ".env",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_capture_status(CaptureStatus::Omitted)
                .with_name("OPENAI_API_KEY")
                .with_metadata(serde_json::json!({"secretLike": true})),
        ),
    ];

    let findings = audit_evidence(&evidence, &build_graph(&evidence));
    let codes: Vec<_> = findings.iter().map(|f| f.code.as_str()).collect();
    assert_eq!(
        codes,
        vec![
            "PARSE_FAILED",
            "PERMISSION_WILDCARD_ADDED",
            "PROJECT_OVERRIDES_USER_POLICY",
            "SYMLINK_SKIPPED",
            "SECRET_LIKE_VALUE_OMITTED"
        ]
    );
    for finding in &findings {
        assert!(!finding.problem.is_empty());
        assert!(!finding.cause.is_empty());
        assert!(!finding.fix.is_empty());
        assert!(finding.path.is_some());
        assert!(finding.evidence_id.is_some());
    }
}

#[test]
fn diffs_semantic_mcp_changes_and_raw_source_changes() {
    let baseline_graph = build_graph(&[item(
        "github-old",
        EvidenceKind::McpServer,
        ".mcp.json",
        EvidenceScope::Project,
        40,
        ItemOverrides::default()
            .with_name("github")
            .with_value(serde_json::json!({
                "transport": "stdio",
                "command": "mcp-github",
                "args": ["--read-only"]
            }))
            .with_checksum("old"),
    )]);
    let current_graph = build_graph(&[item(
        "github-new",
        EvidenceKind::McpServer,
        ".mcp.json",
        EvidenceScope::Project,
        40,
        ItemOverrides::default()
            .with_name("github")
            .with_value(serde_json::json!({
                "transport": "http",
                "url": "https://mcp.example.com/github"
            }))
            .with_checksum("new"),
    )]);

    let diff = diff_graphs(&baseline_graph, &current_graph);
    assert_eq!(diff.raw_source_changes.len(), 1);
    assert_eq!(diff.raw_source_changes[0].source_path, ".mcp.json");
    assert_eq!(diff.raw_source_changes[0].before_evidence_id.as_deref(), Some("github-old"));
    assert_eq!(diff.raw_source_changes[0].after_evidence_id.as_deref(), Some("github-new"));
    assert_eq!(diff.raw_source_changes[0].status, "changed");
    assert_eq!(diff.semantic_changes.len(), 1);
    assert_eq!(diff.semantic_changes[0].code, SemanticChangeCode::McpChanged);
    assert_eq!(diff.semantic_changes[0].entity_name, "github");
    let mut fields = diff.semantic_changes[0].details.changed_fields.clone();
    fields.sort();
    assert_eq!(fields, vec!["command", "transport", "urlHost"]);
}

#[test]
fn diffs_setup_inventory_changes_for_save_titles() {
    let baseline_graph = build_graph(&[
        item(
            "instructions-old",
            EvidenceKind::AgentInstruction,
            "AGENTS.md",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("AGENTS.md")
                .with_value(serde_json::json!({"checksum": "old"}))
                .with_checksum("old"),
        ),
        item(
            "skill-old",
            EvidenceKind::Skill,
            ".claude/skills/legacy-review/SKILL.md",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("legacy-review")
                .with_value(serde_json::json!({"installed": true}))
                .with_checksum("legacy"),
        ),
        item(
            "hook-old",
            EvidenceKind::Hook,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("pre-tool-use")
                .with_value(serde_json::json!({"command": "old-hook"}))
                .with_checksum("hook-old"),
        ),
        item(
            "permission-old",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(git status)"}))
                .with_checksum("permission-old"),
        ),
        item(
            "permission-removed",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(bun test)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(bun test)"}))
                .with_checksum("permission-removed"),
        ),
    ]);
    let current_graph = build_graph(&[
        item(
            "instructions-new",
            EvidenceKind::AgentInstruction,
            "AGENTS.md",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("AGENTS.md")
                .with_value(serde_json::json!({"checksum": "new"}))
                .with_checksum("new"),
        ),
        item(
            "skill-new",
            EvidenceKind::Skill,
            ".claude/skills/react-review/SKILL.md",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("react-review")
                .with_value(serde_json::json!({"installed": true}))
                .with_checksum("skill"),
        ),
        item(
            "hook-new",
            EvidenceKind::Hook,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("pre-tool-use")
                .with_value(serde_json::json!({"command": "new-hook"}))
                .with_checksum("hook-new"),
        ),
        item(
            "hook-added",
            EvidenceKind::Hook,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("post-tool-use")
                .with_value(serde_json::json!({"command": "notify"}))
                .with_checksum("hook-added"),
        ),
        item(
            "permission-new",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(git status)")
                .with_value(serde_json::json!({"action": "deny", "rule": "Bash(git status)"}))
                .with_checksum("permission-new"),
        ),
        item(
            "permission-added",
            EvidenceKind::Permission,
            ".claude/settings.json",
            EvidenceScope::Project,
            40,
            ItemOverrides::default()
                .with_name("Bash(bun run build)")
                .with_value(serde_json::json!({"action": "allow", "rule": "Bash(bun run build)"}))
                .with_checksum("permission-added"),
        ),
    ]);

    let diff = diff_graphs(&baseline_graph, &current_graph);
    let mut codes: Vec<_> = diff
        .semantic_changes
        .iter()
        .map(|change| change.code.as_str())
        .collect();
    codes.sort();
    assert_eq!(
        codes,
        vec![
            "HOOK_ADDED",
            "HOOK_CHANGED",
            "INSTRUCTION_CHANGED",
            "PERMISSION_CHANGED",
            "PERMISSION_CHANGED",
            "PERMISSION_CHANGED",
            "SKILL_ADDED",
            "SKILL_REMOVED"
        ]
    );
    assert!(diff.semantic_changes.iter().any(|change| {
        change.code == SemanticChangeCode::PermissionChanged
            && change
                .details
                .extra
                .get("removed")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
    }));
}